test/challtestsrv/cmd/challtestsrv/main.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/test/challtestsrv"
+)
+
+// managementServer is a small HTTP server that can control a challenge server,
+// adding and deleting challenge responses as required
+type managementServer struct {
+	// A managementServer is a http.Server
+	*http.Server
+	log *log.Logger
+	// The challenge server that is under control by the management server
+	challSrv *challtestsrv.ChallSrv
+	// Shutdown is a channel used to request the management server cleanly shut down
+	shutdown chan bool
+}
+
+// Run runs the management server blocking until a shutdown request is received
+// on the shutdown channel. When a shutdown occurs the management HTTP server
+// will be cleanly shutdown and the provided WaitGroup will have its `Done()`
+// function called. This allows the caller to wait on the waitgroup and know
+// that they will not unblock until the management server is cleanly stopped.
+func (srv *managementServer) Run(wg *sync.WaitGroup) {
+	srv.log.Printf("Starting management server")
+	// Start the HTTP server in its own dedicated Go routine
+	go func() {
+		err := srv.ListenAndServe()
+		if err != nil {
+			srv.log.Print(err)
+		}
+	}()
+
+	// Block forever waiting for a shutdown request
+	<-srv.shutdown
```

Comment:
I think the management server can also follow the pattern of passing through Shutdown requests to its underlying `http.Server`, instead of needing a waitgroup and a channel.